solutions/sbom-chatbot/KnowledgeGraphEnhancedRAG.py

```diff
@@ -0,0 +1,212 @@
+    NeptuneDatabasePropertyGraphStore,
+    NeptuneDatabaseGraphStore,
+)
+from llama_index.llms.bedrock import Bedrock
+from llama_index.embeddings.bedrock import BedrockEmbedding
+from llama_index.core import (
+    PropertyGraphIndex,
+    KnowledgeGraphIndex,
+    VectorStoreIndex,
+    StorageContext,
+    load_index_from_storage,
+    PromptTemplate,
+)
+
+PERSIST_DIR = "persist/"
+
+logger = logging.getLogger(__name__)
+
+
+class KnowledgeGraphEnhancedRAG:
+    """This class shows how to run a PropertyGraphIndex query over data using LlamaIndex"""
+
+    def __init__(
+        self,
+        graph_store: NeptuneDatabasePropertyGraphStore,
+        kg_graph_store: NeptuneDatabaseGraphStore,
+        llm: Bedrock,
+        embed_model: BedrockEmbedding,
+    ):
+        self.graph_store = graph_store
+        self.kg_graph_store = kg_graph_store
+        self.llm = llm
+        self.embed_model = embed_model
+        self._load_kg_index()
+        self._load_vector_index()
+        self.kg_query_engine = self.kg_index.as_query_engine(
+            include_text=False,
+            llm=llm,
+        )
+        self.vector_query_engine = self.vector_index.as_query_engine(
+            llm=llm,
+        )
+        self.vector_retriever = self.vector_index.as_retriever()
+
+    def _load_kg_index(self) -> None:
+        """Creates or loads the KG Index
+
+        Returns:
+            The loaded KG Index
+        """
+        # check if kg storage already exists
+        if not os.path.exists(PERSIST_DIR + "/kg"):
+            # load the documents and create the index
+            logger.info("Creating KnowledgeGraphIndex from documents")
+            kg_storage_context = StorageContext.from_defaults(
+                graph_store=self.kg_graph_store
+            )
+            parser = PDFReader()
+            file_extractor = {".pdf": parser}
+            reader = SimpleDirectoryReader(
+                input_dir="data/kg_enhanced_rag", file_extractor=file_extractor
+            )
+            documents = reader.load_data()
+
+            text = (
+                "Your task is to take the text provided and extract up to the "
+                "{max_knowledge_triplets} most important concepts in "
+                "knowledge triplets in the form of (subject, predicate, object).\n"
+                "Triplets should be focused on entities such as people, companies, locations, and events.\n"
+                "All triplets should not include stopwords such as a, an, and, are, as, at, be, but, by, for, if, in, into, is, it, no, not, of, on, or, such, that, the, their, then, there, these, they, this, to, was, will and with.\n"
+                "For each predicate, simplify the action into a single verb or a 2 words.\n"
+                "---------------------\n"
+                "Example:"
+                "Text: Amazon is located in Seattle."
+                "Triplets:\n(amazon, is located in, seattle)\n"
```

Comment:
Your example here as "is" and "in", which according to the prompt above should not be included.